### DISCUSSION TITLE
Specify custom CA Cert?

### DISCUSSION THREAD

**Discussion by DnOberon** [erlef/oidcc#276] [Q&A] [answered]
I'm still new to Elixir/Erlang and am just wondering how I can specify a custom CA cert to be used in the requests - not finding info on that anywhere :\

**Comment by maennchen** [comment, marked as answer]
This is possible by passing an `ssl` option to the `request_options` of most function.

Example:

* Erlang: `oidcc:retrieve_token(AuthCode, Provider, CLientId, ClientSecret, #{request_opts => #{ssl => [{cacerts, YourCert}}`
* Elixir: `Oidcc.retrieve_token(auth_code, provider, client_id, client_secret, %{request_opts: %{ssl: [cacerts: your_cert]}})`

The `request_opts` are documented here: https://hexdocs.pm/oidcc/oidcc_http_util.html#t:request_opts/0

**Reply by DnOberon** [reply]
How the crap did I miss that, thank you so much

**Reply by maennchen** [reply]
No worries, happens to all of us :) 

The specs are also a bit complicated in some parts and it's not that easy to find the correct place.
I'm also always happy about doc PRs if you think stuff could be explained more clearly.

**Reply by DnOberon** [reply]
So just a quick question, do I need to provide all the ssl opts or can I just provide cacerts?

**Reply by maennchen** [reply]
@DnOberon You'll have to provide complete options. `httpc` provides default ssl options, but it does not merge them.

I would recommend using [`httpc:ssl_verify_host_options/1`](https://www.erlang.org/doc/man/httpc#ssl_verify_host_options-1)

```elixir
ssl_opts = httpc:ssl_verify_host_options(true)
|> Keyword.put(:cacerts, your_cert)
```

This will set `verify` and `customize_hostname_check` for you.

For more details on the option & security hardening, see:
https://erlef.github.io/security-wg/secure_coding_and_deployment_hardening/ssl

**Reply by DnOberon** [reply]
Turns out I needed to use certfile and then needed to drop the default ca certs

```
ssl_opts =
      :httpc.ssl_verify_host_options(true)
      |> Keyword.put(:cacertfile, "/etc/ssl/MYCERTcer")
      |> Keyword.drop([:cacerts])

```

**Reply by DnOberon** [reply]
Well I finally got something working, but it looks like Auth0 adds an extra slash to the issuer and it breaks everything lol

```
Oidcc.ProviderConfiguration.load_configuration(
  "https://dev-KEY.us.auth0.com",
  %{request_opts: %{ssl: ssl_opts}}
) #=> {:error, {:issuer_mismatch, "https://dev-KEY.us.auth0.com/"}}

```

**Reply by DnOberon** [reply]
Oh and I stood up a Keycloak instance to do my testing instead, but it looks like I can't do it unless it's https
```
{:error, {:invalid_config_property, {:uri_https, :userinfo_endpoint}}}
```

**Reply by maennchen** [reply]
@DnOberon

> and it breaks everything lol

Oops :S

The error for `auth0` is correct, the issuer has to match the one from the configuration.
However, passing it in with the slash at the end currently produces a 404 error. This has been solved in #277.

> but it looks like I can't do it unless it's https

`https` is required for the `userinfo_endpoint`. This is per the spec:

> URL of the OP's UserInfo Endpoint [[OpenID.Core]](https://openid.net/specs/openid-connect-discovery-1_0.html#OpenID.Core). **This URL MUST use the https scheme** and MAY contain port, path, and query parameter components.

https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata

There's a lot of providers which do not fully conform to the specs. I'm currently thinking if it makes sense to introduce some kind of `quicks` config param, where you can allow things like ignoring an https requirement or an issuer equality check.

**Reply by DnOberon** [reply]
That was ridiculously fast! 

I like the idea of being able to ignore at least the HTTPS requirement, might help those doing local development against a local IdP like Keycloak or something like that.

**Reply by maennchen** [reply]
@DnOberon It will ve at least a week until I have time to have a look at that. I’m working against a tight deadline in my job.

In the meantime you could use something like localtunnel or ngrok to get a HTTPS url for your local setup.

**Reply by maennchen** [reply]
See #278

**Reply by DnOberon** [reply]
No worries. I'll look at how I can potentially contribute. I'm still new, but this might be simple enough for me to handle. Unfortunately both those tools are locked down by my workplace. I've got another option though.

**Reply by maennchen** [reply]
@DnOberon Release `v3.1.0-beta.1` has a new `quirks` option that allows disabling the `HTTPS` check.

https://hexdocs.pm/oidcc/3.1.0-beta.1/oidcc_provider_configuration.html#t:quirks/0

Does this fit your needs for local development?

**Reply by DnOberon** [reply]
@maennchen tested and it covers my use case. Thank you so much for taking some time to do this, we at the Idaho National Laboratory thank you profusely!

**Reply by maennchen** [reply]
@DnOberon I forgot to ask: I added your provider in the integration test suite:

https://github.com/erlef/oidcc/blob/f00b016e30ad324bb5c49e96cba296f8bff9e9e5/test/oidcc_authorization_SUITE.erl#L13

Is that ok for you?

**Comment by DnOberon** [comment]
Absolutely! Thank you!

**Comment by DnOberon** [comment]
Yeah that’s fine. It’s a public endpoint.
